src/adapters/axelar/AxelarReceiverAdapter.sol

```diff
@@ -0,0 +1,123 @@
+    address public senderAdapter;
+    string public senderChain;
+
+    mapping(bytes32 => bool) public isMessageExecuted;
+    mapping(bytes32 => bool) public commandIdStatus;
+
+    mapping(string => uint256) public reverseChainIdMap;
+
+    /*/////////////////////////////////////////////////////////////////
+                                 MODIFIER
+    ////////////////////////////////////////////////////////////////*/
+    modifier onlyCaller() {
+        if (!gac.isPrevilagedCaller(msg.sender)) {
+            revert Error.INVALID_PREVILAGED_CALLER();
+        }
+        _;
+    }
+
+    /*/////////////////////////////////////////////////////////////////
+                         CONSTRUCTOR
+    ////////////////////////////////////////////////////////////////*/
+    constructor(address _gateway, address _gac) {
+        gateway = IAxelarGateway(_gateway);
+        gac = IGAC(_gac);
+    }
+
+    /*/////////////////////////////////////////////////////////////////
+                         EXTERNAL FUNCTIONS
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IBridgeReceiverAdapter
+    function updateSenderAdapter(bytes memory _senderChain, address _senderAdapter) external override onlyCaller {
+        string memory _senderChainDecoded = abi.decode(_senderChain, (string));
+
+        if (keccak256(abi.encode(_senderChainDecoded)) == keccak256(abi.encode(""))) {
+            revert Error.ZERO_CHAIN_ID();
+        }
+
+        if (_senderAdapter == address(0)) {
+            revert Error.ZERO_ADDRESS_INPUT();
+        }
+
+        address oldAdapter = senderAdapter;
+        senderAdapter = _senderAdapter;
+        senderChain = _senderChainDecoded;
+
+        emit SenderAdapterUpdated(oldAdapter, _senderAdapter, _senderChain);
+    }
+
+    /// @dev accepts new cross-chain messages from axelar gateway
+    function execute(
+        bytes32 commandId,
+        string calldata sourceChain,
+        string calldata sourceAddress,
+        bytes calldata payload
+    ) external override {
+        /// @dev step-1: validate incoming chain id
+        if (keccak256(bytes(sourceChain)) != keccak256(bytes(senderChain))) {
+            revert Error.INVALID_SENDER_CHAIN_ID();
+        }
+
+        /// @dev step-2: validate the caller
+        if (!gateway.validateContractCall(commandId, sourceChain, sourceAddress, keccak256(payload))) {
+            revert Error.NOT_APPROVED_BY_GATEWAY();
+        }
+
+        /// @dev step-3: validate the source address
+        if (sourceAddress.toAddress() != senderAdapter) {
+            revert Error.INVALID_SENDER_ADAPTER();
+        }
+
+        /// decode the cross-chain payload
+        AdapterPayload memory decodedPayload = abi.decode(payload, (AdapterPayload));
+        bytes32 msgId = decodedPayload.msgId;
+
+        /// @dev step-4: check for duplicate message
+        if (commandIdStatus[commandId] || isMessageExecuted[msgId]) {
```

Comment:
If I'm not mistaken, according to the [spec](https://github.com/axelarnetwork/cgp-spec/blob/306de046657e59f13dc822d37b2899d846e951a3/cgp-evm-v1.md?plain=1#L59) and [implementation](https://github.com/axelarnetwork/axelar-cgp-solidity/blob/main/contracts/AxelarGateway.sol#L151), `validateContractCall()` (line 88) should already fail if a message is being re-executed. If so, is this separate check (`isMessageExecuted[]`) for whether a message is already executed ever really used? If it is required for more informative error handling should we move it prior to the `validateContractCall()`

Reply:
Right, but presumably we want to actually detect and provide an informative error message if the command has already been executed. At present, we can't do that because line 88 results in a generic message validation failure from the AMB. I suggest we move this check prior to that line.